### PR TITLE
[Fix-14303][Master] Workflow with sub_process task can't be stopped w…

### LINKED ISSUE #14303: [Bug] [master/task] workflow with sub_process task can't be stop when running
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar issues.

### What happened

version 3.1.7

I want to stop the workflow with 'sub_process' task, but the workflow still 'ready stop' and 'sub_process' task still running.

Task state:
image

Workflow state:
image

'sub_process' task log:
image

---

Task state should be 'kill' instead of 'stop', cause `TaskExecutionStatus` only admit 'kill':

```
  public boolean isFinished() {
      return isSuccess() || isKill() || isFailure() || isPause();
  }

  public boolean isKill() {
      return this == TaskExecutionStatus.KILL;
  }
```

### What you expected to happen

Stop workflow with 'sub_process' task

### How to reproduce

1. create A workflow with a shell task which will sleep 100s.

image

2. create B workflow with a sub_process task which point to A workflow.
3. start B workflow.
4. after few seconds, stop B workflow instance.

### Anything else

_No response_

### Version

3.1.x

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by HomminLee**
## Purpose of the pull request

fix #14303

## Brief change log

When the sub workflow is finished, check the workflow state and transition to task status.

## Verify this pull request

This pull request is already covered by existing tests, such as org.apache.dolphinscheduler.server.master.SubProcessTaskTest#testFinish.

### CODE REVIEW

**Comment by HomminLee** [comment]
@SbloodyS need a label to continue check

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/task/SubTaskProcessor.java [hunk 1]
```diff
@@ -129,14 +129,29 @@ private void updateTaskState() {
...
         }
     }
 
+    private TaskExecutionStatus subProcessFinishedStateTransform(WorkflowExecutionStatus state) {
+        if (state.isSuccess()) {
+            return TaskExecutionStatus.SUCCESS;
+        } else if (state.isFailure()) {
+            return TaskExecutionStatus.FAILURE;
+        } else if (state.isStop()) {
+            return TaskExecutionStatus.KILL;
```
Comment (zhongjiajie) [inline dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/task/SubTaskProcessor.java:145]: why we workflow stop is kill?

Comment (HomminLee) [reply dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/task/SubTaskProcessor.java:145]: STOP is workflow state, and KILL is task state. AFAK, task has no STOP state (`TaskExecutionStatus` has STOP state, but it has not been used anywhere). `TaskExecutionStatus` has code:

```
    public boolean isFinished() {
        return isSuccess() || isKill() || isFailure() || isPause();
    }
```

PAUSE workflow has PAUSE task, only KILL task state can match STOP workflow state, both of these indicate termination.

**Review by caishunfeng — APPROVED** [review]
:+1:

**Comment by caishunfeng** [comment]
The pr target should change to dev branch first?

**Comment by HomminLee** [comment]
> The pr target should change to dev branch first?

In dev branch, sub_process has been refactor #13948 , and there is no relevant code anymore. BTW, I don't know if dev branch have the same issue.

**Comment by HomminLee** [comment]
> > The pr target should change to dev branch first?
> 
> In dev branch, sub_process has been refactor #13948 , and there is no relevant code anymore. BTW, I don't know if dev branch have the same issue.

I browsed sub_process code in dev branch,  the new logic task framework only use three state: RUNNING, SUCCESS, FAILED.(Reference code: `SubWorkflowAsyncTaskExecuteFunction#getAsyncTaskExecutionStatus`) . Should the framework be improved?

**Pull request merged**